src/radio/radio-device-manager.ts

```diff
@@ -0,0 +1,92 @@
+import {RadioController} from './radio-controller';
+import {UdpResponseData} from './dataflow';
+
+/**
+ * The default port to listen on for radio responses.
+ */
+const DEFAULT_LISTEN_PORT = 3311;
+
+/**
+ * An implementation of `smarthome.DeviceManager` that implements radio functionality.
+ */
+export class RadioDeviceManager implements smarthome.DeviceManager {
+  private radioController: RadioController;
+  private listenPort: number;
+  public deviceIdToAddress: Map<string, string> = new Map<string, string>();
+
+  /**
+   * @param  radioController  The radio controller used for radio communication.
+   * @param  listenPort  The port to listen on for radio responses.
+   */
+  constructor(
+    radioController: RadioController,
+    listenPort: number = DEFAULT_LISTEN_PORT
+  ) {
+    this.radioController = radioController;
+    this.listenPort = listenPort;
+  }
+
+  /**
+   * Associates a deviceId with an IP address.
+   * Required for routing Execute commands.
+   * @param deviceId The deviceId of the device
+   * @param address The address of the device
+   */
+  public addDeviceIdToAddress(deviceId: string, address: string): void {
+    this.deviceIdToAddress.set(deviceId, address);
+  }
+
+  markPending(request: smarthome.IntentRequest): Promise<void> {
+    //TODO(cjdaly) implementation
+    return Promise.resolve();
+  }
+
+  getProxyInfo(id: string): smarthome.ProxyInfo {
+    //TODO(cjdaly) implementation
+    throw new Error('Method not implemented.');
+  }
+
+  /**
+   * Fulfills a UdpRquestData.
+   * @param udpRequestData  The UdpRequestData to source radio parameters from.
+   * @returns  A promise that resolves to the determined UdpResponseData.
+   */
+  private async processUdpRequestData(
+    udpRequestData: smarthome.DataFlow.UdpRequestData
+  ): Promise<smarthome.DataFlow.UdpResponseData> {
+    const payload = Buffer.from(udpRequestData.data, 'hex');
+    const localAddress = this.deviceIdToAddress.get(udpRequestData.deviceId);
+    const udpResponse = await this.radioController.sendUdpMessage(
+      payload,
+      localAddress!,
```

Comment:
Do we need to assert it's not undefined? What happens if it is undefined

Reply:
It should be impossible for it to be undefined, since we check that the address is in the map in the `send` call.

(May be unrelated, since I'm not sure if the TypeScript compiler would be able to relate these calls across functions anyways, but there's an open issue where `map.has()` doesn't guarantee `map.get()` is not `undefined`: https://github.com/microsoft/TypeScript/issues/9619)